datafusion/optimizer/src/analyzer/type_coercion.rs

```diff
@@ -103,6 +104,20 @@ fn analyze_internal(
     // select t2.c2 from t1 where t1.c1 in (select t2.c1 from t2 where t2.c2=t1.c3)
     schema.merge(external_schema);
 
+    if let LogicalPlan::Filter(Filter {
```

Comment:
Thanks @xinlifoobar  -- this looks good to me

I am thinking we might be able to avoid the copy -- but we could do it in a follow on PR too

Now that you have written this, I think maybe we could simply call `cast_to`

something like

```rust
if let LogicalPlan::Filter(filter) = &mut plan {
  filter.predicate = filter.predicate.cast_to(&DataType::Boolean, &filter.schema)?;
}

The cast_to is a no op if the source datatype is the the same as the target 🤔

Reply:
Thanks... Sorry, I did not find a way to avoid copying the predicate once, the cast_to returns a result and costumes the predicate..